rpc/send.go

```diff
@@ -0,0 +1,108 @@
+package rpc
+
+import (
+	"context"
+	"unsafe"
+
+	"capnproto.org/go/capnp/v3/internal/mpsc"
+	rpccp "capnproto.org/go/capnp/v3/std/capnp/rpc"
+)
+
+type preparer interface {
+	Prepare(rpccp.Message) error
+}
+
+type sendQueue mpsc.Queue
+
+func newSendQueue() *sendQueue { return (*sendQueue)(mpsc.New()) }
+
+func (sq *sendQueue) Send(ctx context.Context, p preparer) (err error) {
+	cherr := cherrPool.Get()
+	defer cherrPool.Put(cherr)
+
+	sq.SendAsync(ctx, p, cherr)
+
+	select {
+	case err = <-cherr:
+	case <-ctx.Done():
+		err = ctx.Err()
+	}
+
+	return
+}
+
+func (sq *sendQueue) SendAsync(ctx context.Context, p preparer, r ErrorReporter) {
+	(*mpsc.Queue)(sq).Send(sendReq{
+		er:       r,
+		preparer: p,
+	})
+}
+
+func (sq *sendQueue) Recv(ctx context.Context) preparer {
+	v, err := (*mpsc.Queue)(sq).Recv(ctx)
+	if err != nil {
+		return sendFailure(err)
+	}
+
+	return *(*preparer)(unsafe.Pointer(&v))
```

Comment:
I don't think this is correct; the inserted value is a sendReq, and I'm not sure the memory layout will be the same as if it is passed in as a preparer.

You should probably just use a type assertion here anyway:

```go
v.(sendReq)
```

...I do not think `unsafe`is warranted here.